src/Uno.UI/Microsoft/UI/Xaml/Controls/TreeView/TreeViewViewModel.cs

```diff
@@ -0,0 +1,1008 @@
+					offset++;
+					AddNodeToView(childNode, offset + index);
+					offset = AddNodeDescendantsToView(childNode, index, offset);
+				}
+
+				return offset;
+			}
+
+			return offset;
+		}
+
+		private void RemoveNodeAndDescendantsFromView(TreeViewNode value)
+		{
+			if (value.IsExpanded)
+			{
+				int size = value.Children.Count;
+				for (int i = 0; i < size; i++)
+				{
+					var childNode = value.Children[i];
+					RemoveNodeAndDescendantsFromView(childNode);
+				}
+			}
+
+			bool containsValue = IndexOfNode(value, out var valueIndex);
+			if (containsValue)
+			{
+				RemoveAt(valueIndex);
+			}
+		}
+
+		private void RemoveNodesAndDescendentsWithFlatIndexRange(int lowIndex, int highIndex)
+		{
+			if (lowIndex > highIndex)
+			{
+				throw new ArgumentOutOfRangeException(nameof(lowIndex));
+			}
+
+			for (int i = highIndex; i >= lowIndex; i--)
+			{
+				RemoveNodeAndDescendantsFromView(GetNodeAt(i));
+			}
+		}
+
+		private int GetNextIndexInFlatTree(TreeViewNode node)
+		{
+			bool isNodeInFlatList = IndexOfNode(node, out var index);
+
+			if (isNodeInFlatList)
+			{
+				index++;
+			}
+			else
+			{
+				// node is Root node, so next index in flat tree is 0
+				index = 0;
+			}
+			return index;
+		}
+
+		// When ViewModel receives a event, it only includes the sender(parent TreeViewNode) and index.
+		// We can't use sender[index] directly because it is already updated/removed
+		// To find the removed TreeViewNode:
+		//   calcuate allOpenedDescendantsCount in sender[0..index-1] first
+		//   then add offset and finally return TreeViewNode by looking up the flat tree.
+		private TreeViewNode GetRemovedChildTreeViewNodeByIndex(TreeViewNode node, int childIndex)
+		{
+			var allOpenedDescendantsCount = 0;
+			for (var i = 0; i < childIndex; i++)
+			{
+				TreeViewNode calcNode = node.Children[i];
+				if (calcNode.IsExpanded)
+				{
+					allOpenedDescendantsCount += GetExpandedDescendantCount(calcNode);
+				}
+			}
+
+			var childIndexInFlatTree = GetNextIndexInFlatTree(node) + childIndex + allOpenedDescendantsCount;
+			return GetNodeAt(childIndexInFlatTree);
+		}
+
+		private int CountDescendants(TreeViewNode value)
+		{
+			int descendantCount = 0;
+			var size = value.Children.Count;
+			for (var i = 0; i < size; i++)
+			{
+				var childNode = value.Children[i];
+				descendantCount++;
+				if (childNode.IsExpanded)
+				{
+					descendantCount = descendantCount + CountDescendants(childNode);
+				}
+			}
+
+			return descendantCount;
+		}
+
+		private int IndexOfNextSibling(TreeViewNode childNode)
+		{
+			var child = childNode;
+			var parentNode = child.Parent;
+			int stopIndex;
+			bool isLastRelativeChild = true;
+			while (parentNode != null && isLastRelativeChild)
+			{
+				int relativeIndex = parentNode.Children.IndexOf(childNode);
+				if (parentNode.Children.Count - 1 != relativeIndex)
+				{
+					isLastRelativeChild = false;
+				}
+				else
+				{
+					child = parentNode;
```

Comment:
Codacy found an issue: [Remove this useless assignment to local variable 'child'.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=5156819)